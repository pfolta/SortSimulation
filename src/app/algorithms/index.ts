export type { default as Phase } from "@/app/algorithms/Phase";
export * from "@/app/algorithms/Phase";

export type { default as SortingAlgorithm } from "@/app/algorithms/SortingAlgorithm";

export * from "@/app/algorithms/Trace";

export { default as BubbleSort } from "@/app/algorithms/BubbleSort";
