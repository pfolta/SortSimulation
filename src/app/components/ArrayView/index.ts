export { default as ArrayView, ArrayViewProps } from "@/app/components/ArrayView/ArrayView";
export { default as Element, ElementProps, StyledElement } from "@/app/components/ArrayView/Element";
