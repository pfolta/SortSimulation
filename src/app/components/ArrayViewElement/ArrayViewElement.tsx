import styled from "styled-components";

interface ArrayViewElementProps {
    value: number;
}

const ArrayViewElement = styled.div.withConfig({
    shouldForwardProp: (prop) => ["children", "style"].includes(prop)
})<ArrayViewElementProps>`
    background-color: #4c566a;
    color: white;

    transition: transform 100ms ease-out;

    :hover {
        transform: translateY(-1rem);

        /* Prevent jitter on hover, see https://codepen.io/csilverman/post/fixing-the-jitter-bug */
        ::before {
            content: "";

            position: absolute;
            top: 0;
            left: 0;
            right: 0;
            bottom: -1rem;

            z-index: -1;
        }
    }
`;

export default ArrayViewElement;
export { ArrayViewElementProps };
