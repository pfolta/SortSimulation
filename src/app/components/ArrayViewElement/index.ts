export type { ArrayViewElementProps } from "@/app/components/ArrayViewElement/ArrayViewElement";
export { default as ArrayViewElement } from "@/app/components/ArrayViewElement/ArrayViewElement";

export type { ImageElementProps } from "@/app/components/ArrayViewElement/ImageElement";
export { default as ImageElement, StyledImageElement } from "@/app/components/ArrayViewElement/ImageElement";

export type { NumberElementProps } from "@/app/components/ArrayViewElement/NumberElement";
export { default as NumberElement, StyledNumberElement } from "@/app/components/ArrayViewElement/NumberElement";
