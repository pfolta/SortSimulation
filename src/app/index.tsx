import React from "react";
import { render } from "react-dom";
import { setAppElement } from "react-modal";

import App from "@/app/App";

const appElement = document.querySelector("#app") as HTMLElement;

setAppElement(appElement);
render(<App />, appElement);
