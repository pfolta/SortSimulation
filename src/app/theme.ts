import darkTheme from "@/resources/themes/dark.json";
import lightTheme from "@/resources/themes/light.json";

interface Color {
    blue: string;
    gray1: string;
    gray2: string;
    gray3: string;
    gray4: string;
    gray5: string;
    gray6: string;
    green: string;
    orange: string;
    purple: string;
    red: string;
    yellow: string;
}

interface Theme {
    name: string;
    body: {
        background: string;
    };
    colors: Color;
}

const themes = {
    dark: darkTheme,
    light: lightTheme
};

export default Theme;
export { Color, themes };
