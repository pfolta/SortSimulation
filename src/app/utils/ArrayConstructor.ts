import "@/app/utils/Array";

declare global {
    interface ArrayConstructor {
        almostSorted: (size: number, distance: number) => number[];
        fewUnique: (size: number, uniqueCount: number) => number[];
        random: (size: number) => number[];
        range: <T>(start: number | T[], end?: number) => number[];
        reversed: (size: number) => number[];
        sorted: (size: number) => number[];
    }
}

/**
 * Create an array where every element is at most `distance` places away from its sorted position
 */
Array.almostSorted = (size: number, distance: number): number[] =>
    Array.sorted(size)
        .chunked(distance + 1)
        .flatMap((chunk) => chunk.shuffled());

/**
 * Create a randomly shuffled array containing `uniqueCount` unique numbers from 1 to `size`, e.g. [7, 3, 8, 8, 7, 4, 3, 1, 3, 4]
 */
Array.fewUnique = (size: number, uniqueCount: number): number[] => {
    const uniques = Array.random(size).slice(0, uniqueCount);
    return Array.range(size).map(() => uniques[Math.floor(Math.random() * uniques.length)]);
};

/**
 * Create a randomly shuffled array containing all numbers from 1 to `size`, e.g. [3, 5, 1, 2, 4]
 */
Array.random = (size: number): number[] => Array.sorted(size).shuffled();

/**
 * Create a range from `start` to less than `end` (inclusive of `start` but not `end`): [start, end)
 * If only one argument is passed, the range will start at 0: [0, end)
 * If an array is passed, the range will contain the indices of the array provided: [0, array.length)
 */
Array.range = <T>(start: number | T[], end?: number): number[] => {
    if (Array.isArray(start)) {
        return [...start.keys()];
    }

    if (end !== undefined) {
        return Array.from({ length: end - start }, (_, i) => start + i);
    }

    return [...Array(start).keys()];
};

/**
 * Create an array from `size` down to 1 in descending order, e.g. [5, 4, 3, 2, 1]
 */
Array.reversed = (size: number): number[] => Array.sorted(size).reverse();

/**
 * Create a sorted array from 1 to `size`, e.g. [1, 2, 3, 4, 5]
 */
Array.sorted = (size: number): number[] => Array.range(1, size + 1);

export {};
