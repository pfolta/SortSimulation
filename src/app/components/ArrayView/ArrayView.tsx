import React, { useMemo } from "react";
import styled from "styled-components";

import { Element } from "@/app/components/ArrayView";

interface ArrayViewProps {
    array: number[];
}

const Wrapper = styled.div<ArrayViewProps>`
    display: grid;
    grid-template-columns: repeat(${(props) => props.array.length}, 1fr);
    column-gap: 1px;
    align-items: end;

    height: 30vh;
`;

const ArrayView = ({ array }: ArrayViewProps): JSX.Element => {
    const maxValue = useMemo(() => Math.max(...array), [array]);

    return (
        <Wrapper array={array}>
            {array.map((_, index) => (
                <Element key={index} array={array} index={index} maxValue={maxValue} />
            ))}
        </Wrapper>
    );
};

export default ArrayView;
export { ArrayViewProps };
