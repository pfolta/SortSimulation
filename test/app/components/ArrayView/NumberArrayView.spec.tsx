import React from "react";

import renderWithProviders from "../../../utils/renderWithProviders";

import { NumberArrayView } from "@/app/components/ArrayView";
import { StyledNumberElement } from "@/app/components/ArrayViewElement";
import { palette } from "@/app/theme";
import "@/app/utils/ArrayConstructor";

describe("NumberArrayView", () => {
    const array = [4, 2, 5, 1, 3];
    const maxValue = Math.max(...array);
    const colors = Array.range(array).map((index) => palette[index % palette.length]);

    it("renders correctly", () => {
        const component = renderWithProviders(<NumberArrayView array={array} colors={colors} />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders a StyledNumberElement for every array element in the correct order", () => {
        const component = renderWithProviders(<NumberArrayView array={array} colors={colors} />);
        const elements = component.root.findAllByType(StyledNumberElement);

        expect(elements).toHaveLength(array.length);

        elements.forEach((element, index) => {
            expect(element.props.value).toBe(array[index]);
            expect(element.props.maxValue).toBe(maxValue);
        });
    });
});
