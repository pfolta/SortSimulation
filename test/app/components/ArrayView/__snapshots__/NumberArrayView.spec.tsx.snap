// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberArrayView renders correctly 1`] = `
.c0 {
  display: grid;
  grid-template-columns: repeat(5,1fr);
  -webkit-column-gap: 1px;
  column-gap: 1px;
  -webkit-align-items: end;
  -webkit-box-align: end;
  -ms-flex-align: end;
  align-items: end;
  height: 30vh;
}

.c1 {
  background-color: #4c566a;
  color: white;
  -webkit-transition: -webkit-transform 100ms ease-out;
  -webkit-transition: transform 100ms ease-out;
  transition: transform 100ms ease-out;
}

.c1:hover {
  -webkit-transform: translateY(-1rem);
  -ms-transform: translateY(-1rem);
  transform: translateY(-1rem);
}

.c1:hover::before {
  content: "";
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: -1rem;
  z-index: -1;
}

.c2 {
  font-size: 0;
  text-align: center;
}

.c3 {
  display: block;
  height: 100%;
  overflow: hidden;
}

<div
  className="c0"
>
  <div
    className="c1 c2"
    style={
      Object {
        "fontSize": "0px",
        "height": "80%",
      }
    }
  >
    <span
      className="c3"
    >
      4
    </span>
  </div>
  <div
    className="c1 c2"
    style={
      Object {
        "fontSize": "0px",
        "height": "40%",
      }
    }
  >
    <span
      className="c3"
    >
      2
    </span>
  </div>
  <div
    className="c1 c2"
    style={
      Object {
        "fontSize": "0px",
        "height": "100%",
      }
    }
  >
    <span
      className="c3"
    >
      5
    </span>
  </div>
  <div
    className="c1 c2"
    style={
      Object {
        "fontSize": "0px",
        "height": "20%",
      }
    }
  >
    <span
      className="c3"
    >
      1
    </span>
  </div>
  <div
    className="c1 c2"
    style={
      Object {
        "fontSize": "0px",
        "height": "60%",
      }
    }
  >
    <span
      className="c3"
    >
      3
    </span>
  </div>
</div>
`;
